src/clipping/weileratherton.jl

```diff
@@ -0,0 +1,408 @@
+nextvertex(v::RingVertex{Normal}) = v.right
+nextvertex(v::RingVertex{Entering}) = v.right
+nextvertex(v::RingVertex{Exiting}) = v.left
+
+function collectclipped(entering::Vector{RingVertex{Entering}})
+  rings::Vector{Ring} = []
+  visited::Vector{RingVertex} = []
+  for i in eachindex(entering)
+    if entering[i] in visited
+      continue
+    end
+
+    ring::Vector{RingVertex} = []
+    vertex = entering[i]
+    while !(vertex in ring)
+      if vertex in visited
+        break
+      end
+
+      push!(ring, vertex)
+      push!(visited, vertex)
+      vertex = nextvertex(vertex)
+    end
+
+    # Remove duplicates.
+    newring::Vector{RingVertex} = [ring[1]]
+    for i in 2:length(ring)
+      if !isapprox(ring[i].point, newring[end].point)
+        push!(newring, ring[i])
+      end
+    end
+    ring = newring
+
+    # Polygon might start several vertices after the first collected.
+    # This generally happens when there are overlapping edges that lead
+    # to several entering vertices without exiting in between. Then, the
+    # actual polygon is found by discarding the extra vertices before the
+    # proper loop.
+    k = findfirst(x -> ring[end].point == x.point, ring[1:(end - 1)])
+    if !isnothing(k)
+      ring = ring[(k + 1):end]
+    end
+
+    if length(ring) > 2
+      push!(rings, Ring([r.point for r in ring]...))
+    end
+  end
+  return rings
+end
+
+function decide_direction(clipped₁, clipped₂, clipping₁, clipping₂)
+  # The input segments should all share one common vertex.
+  # Form vectors from the common vertex to other vertices.
+  a = minimum(clipped₁) - maximum(clipped₁)
+  b = maximum(clipped₂) - minimum(clipped₂)
+  c = minimum(clipping₁) - maximum(clipping₁)
+  d = maximum(clipping₂) - minimum(clipping₂)
+
+  α = mod(atan(a[2], a[1]), 2pi) # Angle formed by the clipped₁ segment and a horizontal line.
+  β = mod(atan(b[2], b[1]) - α, 2pi) # Angle between clipped₁ and clipped₂.
+  γ = mod(atan(c[2], c[1]) - α, 2pi) # Angle between clipped₁ and clipping₁.
+  δ = mod(atan(d[2], d[1]) - α, 2pi) # Angle between clipped₁ and clipping₂.
+
+  if isapprox(γ, 0.0) || isapprox(γ, 2pi)
+    if δ < β < 2pi
+      return Entering
+    else
+      return Exiting
+    end
+  elseif isapprox(δ, 0.0) || isapprox(δ, 2pi)
+    if γ < β < 2pi
+      return Exiting
+    else
+      return Entering
+    end
+  end
+  if γ < δ && (γ < β || isapprox(β, γ)) && (isapprox(β, δ) || β < δ)
+    return Exiting
+  elseif δ < γ && (δ < β || isapprox(β, δ)) && (isapprox(β, γ) || β < γ)
+    return Entering
+  end
+  return nothing
+end
+
+function insertintersections!(clipping, clipped, point, vtype, entering)
+  intersection = RingVertex{vtype}(point)
+  insertintersection!(clipping, intersection, :left)
+  insertintersection!(clipped, intersection, :right)
+
+  if vtype == Entering
+    push!(entering, intersection)
+  end
+end
+
+function clip(poly::Polygon, ring::Ring, _::WeilerAthertonClipping)
```

Comment:
let's move this function to be the first function in the file for improved clarity

Reply:
Done, but the functions related to the RingVertex data type are still above, close to the data type definition. Some other function reordering also done.